File: packages/mens_circle/Classes/Domain/Model/Event.php
<?php

declare(strict_types=1);

namespace BeardCoder\MensCircle\Domain\Model;

use DateTime;
use TYPO3\CMS\Extbase\Domain\Model\FileReference;
use TYPO3\CMS\Extbase\DomainObject\AbstractEntity;
use TYPO3\CMS\Extbase\Persistence\ObjectStorage;

final class Event extends AbstractEntity
{
    public string $title = '';

    public string $slug = '';

    public string $teaser = '';

    public string $description = '';

    public ?DateTime $eventDate = null;

    public ?DateTime $startTime = null;

    public ?DateTime $endTime = null;

    public string $location = '';

    public string $street = '';

    public string $postalCode = '';

    public string $city = '';

    public string $locationDetails = '';

    public int $maxParticipants = 20 {
        set(int $value) => max(1, $value);
    }

    public string $costBasis = '';

    public bool $isPublished = false;

    public ?FileReference $image = null;

    /**
     * @var ObjectStorage<Registration>
     */
    public ObjectStorage $registrations;

    public function __construct()
    {
        $this->registrations = new ObjectStorage();
    }

    public function addRegistration(Registration $registration): void
    {
        $this->registrations->attach($registration);
    }

    public function removeRegistration(Registration $registration): void
    {
        $this->registrations->detach($registration);
    }

    public function isPast(): bool
    {
        if ($this->eventDate === null) {
            return false;
        }

        $endOfDay = (clone $this->eventDate)->setTime(23, 59, 59);

        return $endOfDay < new DateTime();
    }

    public function getFullAddress(): string
    {
        $parts = array_filter([
            trim($this->street),
            trim($this->postalCode . ' ' . $this->city),
        ]);

        if ($parts === []) {
            return trim($this->location);
        }

        return implode(', ', $parts);
    }

    public function generateIcalContent(string $siteDomain = 'mens-circle.local'): string
    {
        $eventDate = $this->eventDate instanceof DateTime ? clone $this->eventDate : new DateTime();
        $start = clone $eventDate;
        $end = clone $eventDate;

        if ($this->startTime instanceof DateTime) {
            $start->setTime((int)$this->startTime->format('H'), (int)$this->startTime->format('i'), 0);
        }

        if ($this->endTime instanceof DateTime) {
            $end->setTime((int)$this->endTime->format('H'), (int)$this->endTime->format('i'), 0);
        } else {
            $end = (clone $start)->modify('+2 hours');
        }

        $uid = \sprintf('%s@%s', $this->uid, $siteDomain);
        $description = preg_replace('/\s+/', ' ', strip_tags($this->description)) ?: '';

        return implode("\r\n", [
            'BEGIN:VCALENDAR',
            'VERSION:2.0',
            'PRODID:-//Mens Circle//TYPO3//DE',
            'BEGIN:VEVENT',
            'UID:' . $uid,
            'DTSTAMP:' . gmdate('Ymd\\THis\\Z'),
            'DTSTART:' . $start->format('Ymd\\THis'),
            'DTEND:' . $end->format('Ymd\\THis'),
            'SUMMARY:' . $this->title,
            'DESCRIPTION:' . $description,
            'LOCATION:' . $this->getFullAddress(),
            'END:VEVENT',
            'END:VCALENDAR',
            '',
        ]);
    }
}

File: internal/models/event.go
package models

import (
	"fmt"
	"strings"
	"time"
)

// Event is a men's circle gathering that participants can register for.
type Event struct {
	ID              int64      `json:"id"`
	Title           string     `json:"title"`
	Slug            string     `json:"slug"`
	Description     *string    `json:"description"`
	EventDate       time.Time  `json:"event_date"`
	StartTime       string     `json:"start_time"` // "HH:MM"
	EndTime         string     `json:"end_time"`   // "HH:MM"
	Location        string     `json:"location"`
	LocationDetails *string    `json:"location_details"`
	Street          *string    `json:"street"`
	PostalCode      *string    `json:"postal_code"`
	City            *string    `json:"city"`
	Latitude        *float64   `json:"latitude"`
	Longitude       *float64   `json:"longitude"`
	MaxParticipants int        `json:"max_participants"`
	CostBasis       string     `json:"cost_basis"`
	Image           *string    `json:"image"`
	IsPublished     bool       `json:"is_published"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`
	DeletedAt       *time.Time `json:"deleted_at"`

	// ActiveRegistrations is populated by the repository (count of registered
	// or attended registrations). It is not a database column.
	ActiveRegistrations int `json:"active_registrations"`
}

// AvailableSpots returns how many seats are still free.
func (e Event) AvailableSpots() int {
	spots := e.MaxParticipants - e.ActiveRegistrations
	if spots < 0 {
		return 0
	}
	return spots
}

// IsFull reports whether the event has no remaining seats.
func (e Event) IsFull() bool {
	return e.AvailableSpots() <= 0
}

// IsPast reports whether the event date lies before today.
func (e Event) IsPast() bool {
	endOfDay := time.Date(e.EventDate.Year(), e.EventDate.Month(), e.EventDate.Day(), 23, 59, 59, 0, e.EventDate.Location())
	return endOfDay.Before(time.Now())
}

// FullAddress returns the formatted street address, or empty if incomplete.
func (e Event) FullAddress() string {
	if e.Street == nil || e.City == nil {
		return ""
	}
	postal := ""
	if e.PostalCode != nil {
		postal = *e.PostalCode
	}
	return strings.TrimSpace(fmt.Sprintf("%s, %s %s", *e.Street, postal, *e.City))
}

// HasCoordinates reports whether both latitude and longitude are set.
func (e Event) HasCoordinates() bool {
	return e.Latitude != nil && e.Longitude != nil
}

// GenerateICalContent renders an RFC 5545 VEVENT for calendar integration.
func (e Event) GenerateICalContent(baseURL string) string {
	stamp := time.Now().UTC().Format("20060102T150405Z")
	start := e.combine(e.StartTime).Format("20060102T150405")
	end := e.combine(e.EndTime).Format("20060102T150405")
	desc := ""
	if e.Description != nil {
		desc = strings.ReplaceAll(*e.Description, "\n", "\\n")
	}
	location := e.Location
	if addr := e.FullAddress(); addr != "" {
		location = addr
	}
	var b strings.Builder
	b.WriteString("BEGIN:VCALENDAR\r\n")
	b.WriteString("VERSION:2.0\r\n")
	b.WriteString("PRODID:-//Maennerkreis Niederbayern//Events//DE\r\n")
	b.WriteString("CALSCALE:GREGORIAN\r\n")
	b.WriteString("BEGIN:VEVENT\r\n")
	fmt.Fprintf(&b, "UID:event-%d@maennerkreis-niederbayern.de\r\n", e.ID)
	fmt.Fprintf(&b, "DTSTAMP:%s\r\n", stamp)
	fmt.Fprintf(&b, "DTSTART:%s\r\n", start)
	fmt.Fprintf(&b, "DTEND:%s\r\n", end)
	fmt.Fprintf(&b, "SUMMARY:%s\r\n", e.Title)
	fmt.Fprintf(&b, "DESCRIPTION:%s\r\n", desc)
	fmt.Fprintf(&b, "LOCATION:%s\r\n", location)
	fmt.Fprintf(&b, "URL:%s/event/%s\r\n", strings.TrimRight(baseURL, "/"), e.Slug)
	b.WriteString("END:VEVENT\r\n")
	b.WriteString("END:VCALENDAR\r\n")
	return b.String()
}

// combine merges the event date with an "HH:MM" time string.
func (e Event) combine(hhmm string) time.Time {
	h, m := 0, 0
	fmt.Sscanf(hhmm, "%d:%d", &h, &m)
	return time.Date(e.EventDate.Year(), e.EventDate.Month(), e.EventDate.Day(), h, m, 0, 0, e.EventDate.Location())
}
